File: scripts/prepare-product-images.py
import json
import re
from io import BytesIO
from pathlib import Path

import requests
from PIL import Image, UnidentifiedImageError

root = Path(__file__).resolve().parents[1]
catalog_path = root / 'content' / 'valenza-catalog.json'
out_dir = root / 'public' / 'products'

with catalog_path.open('r', encoding='utf-8') as f:
    catalog = json.load(f)

products = catalog.get('products', [])

for product in products:
    category = re.sub(r'[^a-z0-9]+', '-', (product.get('category') or 'uncategorized').lower()).strip('-')
    (out_dir / category).mkdir(parents=True, exist_ok=True)


def slugify(text: str) -> str:
    return re.sub(r'[^a-z0-9]+', '-', (text or '').strip().lower()).strip('-')


def candidate_urls(product: dict):
    urls = []
    for key in ('image', 'imageMedium', 'imageThumb'):
        value = product.get(key)
        if isinstance(value, str) and value.strip():
            urls.append(value.strip())
    for value in product.get('images', []) or []:
        if isinstance(value, str) and value.strip():
            urls.append(value.strip())
    return urls


def download_image(source_url: str, dest_path: Path) -> bool:
    for attempt in range(3):
        try:
            response = requests.get(source_url, timeout=20)
            response.raise_for_status()
            break
        except Exception:
            if attempt == 2:
                return False

    try:
        with Image.open(BytesIO(response.content)) as img:
            img = img.convert('RGB')
            max_side = 1600
            img.thumbnail((max_side, max_side), Image.Resampling.LANCZOS)
            img.save(dest_path, format='WEBP', quality=82, optimize=True)
    except (UnidentifiedImageError, OSError, ValueError):
        return False

    return dest_path.exists()


count = 0
for product in products:
    slug = slugify(product.get('name') or product.get('slug') or 'product')
    category = re.sub(r'[^a-z0-9]+', '-', (product.get('category') or 'uncategorized').lower()).strip('-')
    dest_path = out_dir / category / f'{slug}.webp'
    local_path = f'/products/{category}/{slug}.webp'

    if dest_path.exists():
        product['image'] = local_path
        product['imageThumb'] = local_path
        product['imageMedium'] = local_path
        product['images'] = [local_path]
        continue

    for url in candidate_urls(product):
        if not url.startswith('http'):
            continue
        if download_image(url, dest_path):
            product['image'] = local_path
            product['imageThumb'] = local_path
            product['imageMedium'] = local_path
            product['images'] = [local_path]
            count += 1
            break
        else:
            product['image'] = local_path
            product['imageThumb'] = local_path
            product['imageMedium'] = local_path
            product['images'] = [local_path]

    if not product.get('image') or product['image'] == '':
        product['image'] = local_path
        product['imageThumb'] = local_path
        product['imageMedium'] = local_path
        product['images'] = [local_path]

with catalog_path.open('w', encoding='utf-8') as f:
    json.dump(catalog, f, indent=2, ensure_ascii=False)

print(f'Updated {count} products with local images in {out_dir}')

File: scripts/prepare-product-images.mjs
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';
import { createHash } from 'crypto';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const root = path.resolve(__dirname, '..');
const catalogPath = path.join(root, 'content', 'valenza-catalog.json');
const outDir = path.join(root, 'public', 'products');

const catalog = JSON.parse(fs.readFileSync(catalogPath, 'utf8'));
const products = catalog.products || [];

function slugify(value) {
  return String(value || '')
    .trim()
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, '-')
    .replace(/(^-|-$)/g, '');
}

function categorySlug(value) {
  return slugify(value || 'uncategorized');
}

function ensureDir(dir) {
  fs.mkdirSync(dir, { recursive: true });
}

function candidateUrls(product) {
  const urls = [];
  for (const key of ['image', 'imageMedium', 'imageThumb']) {
    const value = product[key];
    if (typeof value === 'string' && value.trim()) urls.push(value.trim());
  }
  for (const value of product.images || []) {
    if (typeof value === 'string' && value.trim()) urls.push(value.trim());
  }
  return [...new Set(urls)];
}

async function downloadImage(url, destPath) {
  const response = await fetch(url, { headers: { 'User-Agent': 'Mozilla/5.0' } });
  if (!response.ok) return false;
  const arrayBuffer = await response.arrayBuffer();
  fs.writeFileSync(destPath, Buffer.from(arrayBuffer));
  return fs.existsSync(destPath);
}

async function convertToWebp(sourcePath, destPath) {
  const sharp = await import('sharp');
  await sharp.default(sourcePath).resize({ width: 1600, height: 1600, fit: 'inside', withoutEnlargement: true }).webp({ quality: 82, effort: 6 }).toFile(destPath);
  return fs.existsSync(destPath);
}

async function processProduct(product) {
  const slug = slugify(product.name || product.slug || 'product');
  const category = categorySlug(product.category);
  const dir = path.join(outDir, category);
  ensureDir(dir);
  const destPath = path.join(dir, `${slug}.webp`);
  const localPath = `/products/${category}/${slug}.webp`;

  if (fs.existsSync(destPath)) {
    product.image = localPath;
    product.imageThumb = localPath;
    product.imageMedium = localPath;
    product.images = [localPath];
    return true;
  }

  for (const url of candidateUrls(product)) {
    if (!url.startsWith('http')) continue;
    const tempPath = path.join(dir, `${slug}-${createHash('md5').update(url).digest('hex').slice(0,8)}.tmp`);
    try {
      const downloaded = await downloadImage(url, tempPath);
      if (!downloaded) continue;
      const converted = await convertToWebp(tempPath, destPath);
      fs.unlinkSync(tempPath);
      if (converted) {
        product.image = localPath;
        product.imageThumb = localPath;
        product.imageMedium = localPath;
        product.images = [localPath];
        return true;
      }
    } catch {
      try { fs.unlinkSync(tempPath); } catch {}
    }
  }

  product.image = localPath;
  product.imageThumb = localPath;
  product.imageMedium = localPath;
  product.images = [localPath];
  return false;
}

async function main() {
  let count = 0;
  for (const product of products) {
    const ok = await processProduct(product);
    if (ok) count += 1;
  }
  fs.writeFileSync(catalogPath, JSON.stringify(catalog, null, 2));
  console.log(`Updated ${count} products with local image assets.`);
}

main().catch((error) => {
  console.error(error);
  process.exit(1);
});
